File: src/guillotine-swift/Sources/GuillotineEVM/ExecutionResult.swift
import Foundation
import GuillotinePrimitives

/// Result of EVM bytecode execution
public struct ExecutionResult: Sendable {
    /// Whether execution was successful
    public let success: Bool
    
    /// Amount of gas consumed
    public let gasUsed: UInt64
    
    /// Return data from execution
    public let returnData: Bytes
    
    /// Error message (if execution failed)
    public let errorMessage: String?
    
    /// Create execution result
    public init(
        success: Bool,
        gasUsed: UInt64,
        returnData: Bytes = .empty,
        errorMessage: String? = nil
    ) {
        self.success = success
        self.gasUsed = gasUsed
        self.returnData = returnData
        self.errorMessage = errorMessage
    }
    
    /// Check if execution was successful
    public var isSuccess: Bool {
        success
    }
    
    /// Check if execution reverted
    public var isRevert: Bool {
        !success && errorMessage != nil
    }
}

// MARK: - Equatable
extension ExecutionResult: Equatable {
    public static func == (lhs: ExecutionResult, rhs: ExecutionResult) -> Bool {
        lhs.success == rhs.success &&
        lhs.gasUsed == rhs.gasUsed &&
        lhs.returnData == rhs.returnData &&
        lhs.errorMessage == rhs.errorMessage
    }
}

// MARK: - CustomStringConvertible
extension ExecutionResult: CustomStringConvertible {
    public var description: String {
        if success {
            return "ExecutionResult(success: true, gasUsed: \(gasUsed), returnData: \(returnData.hexString))"
        } else {
            let error = errorMessage ?? "unknown error"
            return "ExecutionResult(success: false, gasUsed: \(gasUsed), error: \(error))"
        }
    }
}

File: sdks/go/evm/execution_result.go
package evm

import "github.com/evmts/guillotine/bindings/go/primitives"

// ExecutionResult represents the result of EVM execution
type ExecutionResult struct {
	success      bool
	gasUsed      uint64
	returnData   primitives.Bytes
	revertReason primitives.Bytes
}

// IsSuccess returns true if the execution was successful
func (r *ExecutionResult) IsSuccess() bool {
	return r.success
}

// GasUsed returns the amount of gas used during execution
func (r *ExecutionResult) GasUsed() uint64 {
	return r.gasUsed
}

// ReturnData returns the data returned by the execution
func (r *ExecutionResult) ReturnData() primitives.Bytes {
	return r.returnData
}

// RevertReason returns the revert reason if execution failed
func (r *ExecutionResult) RevertReason() primitives.Bytes {
	return r.revertReason
}

// HasReturnData returns true if there is return data
func (r *ExecutionResult) HasReturnData() bool {
	return !r.returnData.IsEmpty()
}

// HasRevertReason returns true if there is a revert reason
func (r *ExecutionResult) HasRevertReason() bool {
	return !r.revertReason.IsEmpty()
}